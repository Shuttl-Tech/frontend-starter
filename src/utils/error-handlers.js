import { BAD_REQUEST } from 'http-status-codes';
import API from 'apis';

export const handleFailedResponses = (payload: Object, url: string) => {
	const ignoreUrls = [];
	if (!payload.success && ignoreUrls.includes(url)) throw BAD_REQUEST;
	return payload;
}
