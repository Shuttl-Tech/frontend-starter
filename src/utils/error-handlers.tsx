import { AxiosError } from 'axios';
import { INTERNAL_SERVER_ERROR, NOT_FOUND } from 'http-status-codes';
import Toast, { TYPE } from 'utils/toasts';
import * as React from 'react';


export const handleHTTPErrors = (err: AxiosError) => {
  if (!(err.response && err.response.status)) return;
  if (err.response.status === NOT_FOUND) return;

  if (err.response.status === INTERNAL_SERVER_ERROR) {
  	Toast(<div>This request failed because of a SERVER ERROR.<br/>Please check logs, and notify the developers.</div>, { type: TYPE.ERROR, autoClose: false });
  } else if (err.response.data && err.response.data.error && err.response.data.error.message) {
  	Toast(<div>Error:<br/>{err.response.data.error.message}</div>, { type: TYPE.ERROR });
  }
};
