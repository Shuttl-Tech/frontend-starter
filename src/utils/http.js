import axios from 'axios';
import { SCHEMA, normalize, serialize } from 'adapters/application';
import { initializeRequestMocking } from 'mocks/requests';

export const NAMESPACE = "";
export const REQUEST_TIMEOUT = 30000;

const namespacedUrl = (url) => NAMESPACE + url;

const doMockRequests = process.env.REACT_APP_MOCK_REQUESTS === 'true' || process.env.NODE_ENV === 'test';
if (doMockRequests) { initializeRequestMocking(); }

// Functions here are intended to return promises.
// Any async/await or similar handling should be done at a level higher than here.

/**
 * Return `fetch` API resource.
 * @param {String} url
 * @param options
 * @returns {Promise} `Promise<Response>`
 */
export const getRaw = (url: string, options: Object = {}): Promise<Response> => {
	return fetch(namespacedUrl(url), { credentials: 'include', ...options }).then(response => {
		try {
			if (!response.ok)	return Promise.reject({status: response.status, data: response});
		}
		catch (e) {
			// do nothing as such
			console.error(e);
		}
		return response;
	});
};

/**
 * Return `axios` API resource.
 * https://github.com/axios/axios#response-schema
 * @param {String} url
 * @param {Object} options
 * @param {String} schema
 * @returns {Promise} `Promise<Response>`
 */
export const get = async (url: string, options: Object = {}, schema: string = SCHEMA.GENERIC): Promise<Response> => {
	options = modifyRequestOptions(options, schema);
	let response = await axios.request({ url: namespacedUrl(url), method: 'get', ...options });
	return response.data;
};

/**
 * Return `axios` API resource.
 * https://github.com/axios/axios#response-schema
 * @param {String} url
 * @param {Object} options
 * @param {String} schema
 * @returns {Promise} `Promise<Response>`
 */
export const post = async (url: string, options: Object = {}, schema: string = SCHEMA.GENERIC): Promise<Response> => {
	options = modifyRequestOptions(options, schema);
	let response = await axios.request({ url: namespacedUrl(url), method: 'post', ...options });
	return response.data;
};

// Other utility functions
/**
 * Modify all request options, such as headers, mode, credentials, body, etc.
 *
 * For more details on options, see: https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
 * @param {Object} options Has an additional non-standard `formData: Boolean` property, deleted before flight.
 * @param {String} schema
 * @returns {object} `fetch` options
 */
function modifyRequestOptions(options: Object = {}, schema: string = SCHEMA.GENERIC): {} {

	let headers = modifyRequestHeaders(options || {});

	delete options.formData;	// non-standard optional property

	return {...options,
		headers,
		withCredentials: true,
		data: options.body,
		params: options.body,
		transformRequest: [...axios.defaults.transformRequest, (data) => {
			return normalize(data, schema);
		}],
		transformResponse: [...axios.defaults.transformResponse, (data) => {
			return serialize(data, schema);
		}],
		timeout: REQUEST_TIMEOUT
	};
}

/**
 * Modify all request headers.
 *
 * For more details on headers, see: https://developer.mozilla.org/en-US/docs/Web/API/Headers
 * @returns {object} request headers
 * @param options
 */
function modifyRequestHeaders(options: {} = {}): {} {
	let headers;
	if (options.formData) {
		headers = {...options.headers, "Content-Type": "application/x-www-form-urlencoded" };
	}
	return {...headers};
}
