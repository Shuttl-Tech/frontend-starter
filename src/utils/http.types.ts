import { AxiosRequestConfig } from 'axios';

export type CustomOptionsProps = {
  formData?: boolean,
  body?: any,
  headers?: object
};

export type ModifyRequestOptionsProps = AxiosRequestConfig & CustomOptionsProps;
