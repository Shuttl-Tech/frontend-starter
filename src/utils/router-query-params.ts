import { parse, stringify } from 'query-string';

/**
 * @param queryParams
 */
export const removeQueryParams = (...queryParams: string[]) => {
  let { origin, pathname, search } = window.location;
  let history = window.history;

  let params = { ...parse(search) };
  queryParams.map(qp => delete params[qp]);
  let result = stringify(params);

  result = (result.length) ? `?${result}` : '';

  history.replaceState(history.state, document.title, `${origin}${pathname}${result}`)
};
