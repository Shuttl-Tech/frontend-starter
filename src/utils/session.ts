import cookie from 'js-cookie';

export const TOKEN_COOKIE_NAME = 'authToken';

export function refreshSessionToken(token: string) {
	cookie.remove(TOKEN_COOKIE_NAME);
	setSessionToken(token);
}

export function setSessionToken(token: string) {
	cookie.set(TOKEN_COOKIE_NAME, token, { expires: 7 });
}

export function getSessionToken() {
	return cookie.get(TOKEN_COOKIE_NAME);
}
