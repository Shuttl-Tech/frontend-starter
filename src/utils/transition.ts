import { CompositeUrl, ObjectUrl } from 'apis';
import { History } from 'history';

const ROUTE_PARAM_REGEX = /:([a-z](?:\w+)?|\?)+/ig;


const HISTORY_INSTANCE_PROP = 'applicationHistoryInstance';

export const getHistoryInstance = (): History => {
	// @ts-ignore -- too trivial to bother typing this
	return window[HISTORY_INSTANCE_PROP];
};

export const setHistoryInstance = (history: History) => {
	// @ts-ignore -- too trivial to bother typing this
	return window[HISTORY_INSTANCE_PROP] = history;
};

/**
 * Transition to a specified route.
 * Specify route params as arguments.
 * When calling this function, do this: `transitionTo(this, <route>, ...<params>)`
 *
 * Also ensure that the component is exported using `withRouter` from `react-router-dom`.
 * @param {ReactElementLike} ctx
 * @param {string} route
 * @param params
 */
export function transitionTo(route: string, ...params: string[]) : void {
	const history = getHistoryInstance();
	if (!history) throw new Error('Ensure that history instance has been initialised, because it was not.');
	let transitionRoute = parametrizePath(route, ...params);
	history.push(transitionRoute);
}

/**
 * Turn a PARAMETRIZED_ROUTE (or a pattern matched string) into a route with params baked into it.
 * Example: `/path/:param1/subpath/:param2`, with params `hello`, and `world`,
 * will turn into `/path/hello/subpath/world`.
 *
 * If fewer params are passed than what the route can support, the extra route-params are ignored.
 * Example: `/path/:param1/subpath/:param2`, with params `just-this`,
 * will turn into `/path/just-this/subpath/:param2`.
 *
 * If more params are passed than what the route can support, they are ignored
 * and the output is the same as the `/path/hello/subpath/world` example.
 * Ideally, we can just append them to the path after `join`ing them with `/`.
 *
 * @param {ObjectUrl} route
 * @param {arguments} params
 */
type ReplacerParams = {[key: string]: string };

export function parametrizePath(path: string, ...params: Array<string | number> | [ReplacerParams]) : string;
export function parametrizePath(path: CompositeUrl, ...params: Array<string | number> | [ReplacerParams]) : CompositeUrl;
export function parametrizePath(path: ObjectUrl, ...params: Array<string | number> | [ReplacerParams]) : ObjectUrl {
	let stringPath: string = path as string;
	let parametrizedPath = '';
	const pathIsCompositeUrl = typeof path === 'object';

	if (pathIsCompositeUrl) {
		stringPath = (path as CompositeUrl).URL as string;
	}

	if (params[0] && typeof params[0] === 'object') {
		let dictionary: ReplacerParams = params[0];
		parametrizedPath = stringPath.replace(ROUTE_PARAM_REGEX, (match) => dictionary[match.slice(1)] || match);
	} else {
		const stringParams = params as Array<string>;
		let replaceIndex = 0;
		parametrizedPath = stringPath.replace(ROUTE_PARAM_REGEX, (match) => stringParams[replaceIndex++] || match);
	}

	if (pathIsCompositeUrl) {
		return { ...(path as CompositeUrl),
			URL: parametrizedPath
		};
	}

	return parametrizedPath;
}
