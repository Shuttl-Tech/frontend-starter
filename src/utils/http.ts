import axios, { AxiosRequestConfig } from 'axios';
import { normalize, serialize } from 'adapters/application';
import { initializeRequestMocking } from 'mocks/requests';
import { getSessionToken } from 'utils/session';
import { ObjectUrl } from 'apis';
import { handleHTTPErrors } from 'utils/error-handlers';
import { SCHEMA } from 'adapters/adapter.types';
import { CustomOptionsProps, ModifyRequestOptionsProps } from 'utils/http.types';

export const NAMESPACE = 'api';
export const REQUEST_TIMEOUT = 30000;

const VERSION = 'v1';
const BASE = process.env.REACT_APP_API_BASE_URL;

export const namespacedUrl = (url: string) : string => `${BASE}/${NAMESPACE}/${VERSION}${url}`;
export const processedUrl = (url: ObjectUrl) : string => {
  if (typeof url === 'string') {
  	return namespacedUrl(url);
  }
  if (url.OVERRIDE_BASE) return url.URL;
  return namespacedUrl(url.URL);
};

export const doMockRequests = process.env.REACT_APP_MOCK_REQUESTS === 'true' || process.env.NODE_ENV === 'test';
if (doMockRequests) { initializeRequestMocking(); }

const handleRequestCore = async (url: ObjectUrl, options: AxiosRequestConfig) => {
  try {
  	const request = axios({ url: processedUrl(url), ...options });
  	const response = await request;
  	return response.data;
  }
  catch (err) {
  	handleHTTPErrors(err);
  	throw err;
  }
};

// Functions here are intended to return promises.
// Any async/await or similar handling should be done at a level higher than here.

/**
 * Return `fetch` API resource.
 * @param {ObjectUrl} url
 * @param {object} options
 * @returns {Promise<Response>} response
 */
export const getRaw = (url: ObjectUrl, options: object = {}): Promise<Response> => {
  return fetch(processedUrl(url), { credentials: 'include', ...options }).then(response => {
  	try {
  		if (!response.ok)	throw Object({status: response.status, data: response});
  	}
  	catch (e) {
  		// do nothing as such
  		console.error(e);
  	}
  	return response;
  });
};

/**
 * Return `axios` API resource.
 * https://github.com/axios/axios#response-schema
 * @param {ObjectUrl} url
 * @param {object} options
 * @param {SCHEMA} schema
 * @returns {Promise<{ data: T }>} response
 */
export async function get<T>(url: ObjectUrl, options = {}, schema = SCHEMA.GENERIC): Promise<{ data: T }> {
  options = modifyRequestOptions({ ...options, method: 'get' }, schema);
  return await handleRequestCore(url, options);
}
/**
 * Return `axios` API resource.
 * https://github.com/axios/axios#response-schema
 * @param {ObjectUrl} url
 * @param {object} options
 * @param {SCHEMA} schema
 * @returns {Promise<{ data: T }>} response
 */
export async function post<T>(url: ObjectUrl, options = {}, schema = SCHEMA.GENERIC): Promise<{ data: T }> {
  options = modifyRequestOptions({ ...options, method: 'post' }, schema);
  return await handleRequestCore(url, options);
}
/**
 * Return `axios` API resource.
 * https://github.com/axios/axios#response-schema
 * @param {ObjectUrl} url
 * @param {object} options
 * @param {SCHEMA} schema
 * @returns {Promise<{ data: T }>} response
 */
export async function put<T>(url: ObjectUrl, options = {}, schema = SCHEMA.GENERIC): Promise<{ data: T }> {
  options = modifyRequestOptions({ ...options, method: 'put' }, schema);
  return await handleRequestCore(url, options);
}
/**
 * Return `axios` API resource.
 * https://github.com/axios/axios#response-schema
 * @param {ObjectUrl} url
 * @param {object} options
 * @param {SCHEMA} schema
 * @returns {Promise<{ data: T }>} response
 */
export async function patch<T>(url: ObjectUrl, options = {}, schema = SCHEMA.GENERIC): Promise<{ data: T }> {
  options = modifyRequestOptions({ ...options, method: 'patch' }, schema);
  return await handleRequestCore(url, options);
}
// Other utility functions
/**
 * Modify all request options, such as headers, mode, credentials, body, etc.
 *
 * For more details on options, see: https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch
 * @param {object} options Has an additional non-standard `formData: Boolean` property, deleted before flight.
 * @param {SCHEMA} schema
 * @returns {object} `fetch` options
 */
function modifyRequestOptions(options: ModifyRequestOptionsProps = {}, schema: SCHEMA = SCHEMA.GENERIC): AxiosRequestConfig {

  let headers = modifyRequestHeaders(options || {});

  delete options.formData;	// non-standard optional property

  let isGet = options.method === 'get';
  if (isGet) {
  	options = { ...options, params: options.body }
  } else {
  	options = { ...options, data: options.body }
  }
  delete options.body;  // non-standard optional property

  return {
  	headers,
  	withCredentials: true,
  	transformRequest: [...axios.defaults.transformRequest, (data: object) => {
  		return normalize(data, schema, options);
  	}],
  	transformResponse: [...axios.defaults.transformResponse, (data: any) => {
  		return serialize(data, schema, options);
  	}],
  	timeout: REQUEST_TIMEOUT,
  	...options
  };
}

/**
 * Modify all request headers.
 *
 * For more details on headers, see: https://developer.mozilla.org/en-US/docs/Web/API/Headers
 * @returns {object} request headers
 * @param {object} options
 */
function modifyRequestHeaders(options: CustomOptionsProps = {}): {} {
  let headers;
  if (options.formData) {
  	headers = {...options.headers, 'Content-Type': 'application/x-www-form-urlencoded' };
  }
  headers = {...headers,
  	'x-session-id': getSessionToken()
  };
  return {...headers};
}
