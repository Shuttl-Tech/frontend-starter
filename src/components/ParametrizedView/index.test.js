import React from 'react';
import ReactDOM from 'react-dom';
import { Route } from 'react-router-dom';
import renderer from 'react-test-renderer';

import { component as ParametrizedView } from './';
import { parametrizePath } from 'utils/transition';

it('component tree looks like how it should', () => {
	let props = {
		match: {
			params: {
				id: 1
			}
		}
	};

	const component = renderer.create(<ParametrizedView {...props} />);
	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});
