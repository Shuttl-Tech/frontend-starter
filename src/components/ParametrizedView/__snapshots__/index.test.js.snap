// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component tree looks like how it should 1`] = `
<div>
  <span>
    Welcome to view 1
  </span>
  <div
    className={undefined}
  >
    <button
      onClick={[Function]}
    >
      Change to DE
    </button>
    <button
      onClick={[Function]}
    >
      Change to EN
    </button>
  </div>
</div>
`;
